sympy/tensor/riemannian_geometry.py

```diff
@@ -0,0 +1,1857 @@
+        =(diff(g[j,k],x[i])+diff(g[i,k],x[j])-diff(g[i,j],x[k]))/2.
+
+    Examples:
+    =========
+
+    >>> from sympy.tensor.riemannian_geometry import christoffel_1
+    >>> from sympy.tensor.arraypy import Arraypy, TensorArray
+    >>> from sympy import symbols, cos
+    >>> x1, x2 = symbols('x1, x2')
+
+    var is a list of symbolic arguments. May be a list, one-dimensional
+    arraypy or one-dimensional tensor with valence of indices (+1):
+
+    >>> var = [x1, x2]
+    >>> A = Arraypy((2, 2))
+    >>> g = TensorArray(A,(-1, -1))
+    >>> g[0,0] = cos(x2)**2
+    >>> g[0,1] = 0
+    >>> g[1,0] = 0
+    >>> g[1,1] = 1
+
+    type_output it's optional parameter function, indicating the type of calculation
+    result and receiving the character or string value:
+    - symbol 't' means that the type of the result will match TensorArray;
+    - symbol 'a' means that the type of the result will be Arraypy;
+    - default function takes a parameter 't', so that the result will be a TensorArray.
+
+    The Christoffel symbols of the first kind:
+    >>> ch_1 = christoffel_1(g, var, 't')
+    >>> print(ch_1)
+    0  sin(x2)*cos(x2)
+    -sin(x2)*cos(x2)  0
+    -sin(x2)*cos(x2)  0
+    0  0
+    >>> ch_1.type_pq
+    (0, 3)
+
+    """
+    # Handling of input vector of arguments - var
+    check_vector_of_arguments(var)
+
+    if isinstance(var, (TensorArray, Arraypy)):
+        var = var.to_list()
+    # Definition of number of variables
+    n = len(var)
+
+    # Handling of a input argument - metric tensor g
+    check_metric_tensor(g)
+
+    if isinstance(g, (Arraypy, TensorArray)):
+        if not (g.start_index[0] == g.start_index[1]):
+            raise ValueError(
+                'The starting indices of metric tensor must be identical')
+        idx_start = g.start_index[0]
+    elif isinstance(g, Matrix):
+        if not g.is_symmetric():
+            raise ValueError('The metric tensor must be symmetric.')
+        idx_start = 0
+
+    # The definition of diapason changes in an index
+    [n1, n2] = g.shape
+    if not n == n1:
+        raise ValueError(
+            'The rank of the metric tensor does not coincide with the number of variables.')
+
+    indices = range(idx_start, idx_start + n)
+
+    # Creating of output array with new indices
+    Ch = Arraypy([3, n, idx_start])
+
+    # Calculation
+    for i in indices:
+        for j in indices:
+            for k in indices:
+                Ch[i,
+                   j,
+                   k] = (diff(g[j,
+                                k],
+                              var[i - idx_start]) + diff(g[i,
+                                                           k],
+                                                         var[j - idx_start]) - diff(g[i,
+                                                                                      j],
+                                                                                    var[k - idx_start])) / 2
```

Comment:
What about functions performing tensor products and contractions instead of nested for-loops? That would be easier to read.